Accessibility design doc (#195)
servo/book: src/design-documentation/accessibility/tree.md

> # Generating accessibility trees for web content: `layout::accessibility_tree`

Author: tree.md done 13 july

We generate and update the accessibility tree for a single pipeline (i.e. a single web page being rendered) as part of the [reflow](https://doc.servo.org/layout_api/trait.Layout.html#tymethod.reflow) routine.
This allows us to ensure that all layout information is up to date before updating the accessibility tree, that the tree is updated whenever any DOM mutations and/or style recalculations have occurred, and that DOM mutations are prevented from occurring while the tree is being updated.

The logic for generating and updating the accessibility tree lives in [`layout::accessibility_tree`](https://doc.servo.org/layout/accessibility_tree/index.html).
This module includes:
- [`AccessibilityTree`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html), representing the current state of the accessibility tree and the logic to update the tree based on the current document state
- [`AccessibilityNode`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityNode.html), representing a single node in the accessibility tree and the logic to update it based on its corresponding DOM node

> - [`AccessibilityUpdate`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityUpdate.html), representing the in-process update pass, and providing the `TreeUpdate` when the update is complete.

Author: ```suggestion
- [`AccessibilityUpdate`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityUpdate.html), representing the in-progress update pass, and yielding the `TreeUpdate` when the update is complete.
```

Reviewer: Done.

Author: unchanged as of c0cff9ea292a9b62bff225eb444f85e8775c8941?

Reviewer: Indeed, it was in fact not done. 

Now done.

Author: the s/process/progress/ looks good, but i think we still need to s/providing/yielding/

Reviewer: 🤦 This time for sure!

Author: no worries!

> Our goal is to make updating the accessibility tree fast enough that users can't notice any performance degradation when accessibility is active.

Author: ```suggestion
Our aspirational goal is to make updating the accessibility tree fast enough that users won't notice any performance degradation when accessibility is active.
```

i think it would work with or without “aspirational”, i don’t think we’re being dishonest or anything

Reviewer: Done.

I like aspirational:
- It acknowledges that this may indeed not be possible at all, but that we're nevertheless going to keep shooting for as close as we can get to it;
- It hopefully gestures towards acknowledging that every browser has this as an aspirational goal, we're not the first people to think of it :)

## Activating accessibility for a pipeline

When the ScriptThread receives the `ScriptThreadMessage::SetAccessibilityActive` message for a pipeline, it locates the Document for that pipeline, and calls the [`set_accessibility_active()`](https://doc.servo.org/layout_api/trait.Layout.html#tymethod.set_accessibility_active) method on its `LayoutThread`.

This causes the `LayoutThread` to:
- toggle its [`accessibility_active`](https://doc.servo.org/layout/layout_impl/struct.LayoutThread.html#structfield.accessibility_active) flag,
- create an instance of [`AccessibilityTree`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html) which is stored in its [`accessibility_tree`](https://doc.servo.org/layout/layout_impl/struct.LayoutThread.html#structfield.accessibility_tree) field, and
- set its [`needs_accessibility_update`](https://doc.servo.org/layout/layout_impl/struct.LayoutThread.html#structfield.needs_accessibility_update) flag, which marks it as needing the accessibility tree to be updated in the next reflow.

> If no reflow is otherwise required, setting the `needs_accessibility_update` flag will ensure a rendering update occurs, as it is checked in Document's [`needs_rendering_update()`](https://doc.servo.org/script/dom/document/document/struct.Document.html#method.needs_rendering_update) method, and LayoutThread's [`can_skip_reflow_request_entirely()`](https://doc.servo.org/layout/layout_impl/struct.LayoutThread.html#method.can_skip_reflow_request_entirely) method.

Author: ```suggestion
Even if no reflow is otherwise required, setting the `needs_accessibility_update` flag will ensure a rendering update occurs, as it is checked in Document's [`needs_rendering_update()`](https://doc.servo.org/script/dom/document/document/struct.Document.html#method.needs_rendering_update) method, and LayoutThread's [`can_skip_reflow_request_entirely()`](https://doc.servo.org/layout/layout_impl/struct.LayoutThread.html#method.can_skip_reflow_request_entirely) method.
```

consider adding an example of why this is important?

Reviewer: Done.

When the next reflow occurs, if the LayoutThread's `accessibility_tree` exists, the accessibility tree update occurs as a phase after all the other phases in [`handle_reflow()`](https://doc.servo.org/layout/layout_impl/struct.LayoutThread.html#method.handle_reflow).
This primarily consists of calling the [`update_tree()`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#method.update_tree) method on `AccessibilityTree`, which returns an [`accesskit::TreeUpdate`](https://doc.servo.org/accesskit/struct.TreeUpdate.html).

> The `TreeUpdate` is emitted back to the embedder via [`EmbedderMessage::AccessibilityTreeUpdate`](https://doc.servo.org/servo/enum.EmbedderMsg.html#variant.AccessibilityTreeUpdate), where the `WebView` can forward it to the embedding application via [`WebViewDelegate::notify_accessibility_tree_update()`](https://doc.servo.org/servo/trait.WebViewDelegate.html#method.notify_accessibility_tree_update). (See [WebView accessibility internals](webview-accessibility-internals) for more information on how `TreeUpdate`s are processed in `WebView`.)

Author: broken link? `(webview-accessibility-internals)`

Reviewer: Fixed.

## `AccessibilityTree`

> [`AccessibilityTree`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html) represents the state of the accessibility tree as of the last reflow for a particular Document, and encapsulates the logic to update the tree based on the current document state and produce a `accesskit::TreeUpdate` which can be forwarded to embedders.

Author: ```suggestion
[`AccessibilityTree`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html) represents the state of the accessibility tree as of the last reflow for a particular Document, and encapsulates the logic to update the tree based on the current document state and produce an `accesskit::TreeUpdate` which can be forwarded to embedders.
```

Reviewer: Fixed.

### Updating the tree: `AccessibilityTree::update_tree()`

Each time a reflow occurs, `LayoutThread` will check its [`needs_accessibility_update`](https://doc.servo.org/layout/layout_impl/struct.LayoutThread.html#structfield.needs_accessibility_update) flag, and if it is set will call the [`update_tree()`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#method.update_tree) method on its [`accessibility_tree`](https://doc.servo.org/layout/layout_impl/struct.LayoutThread.html#structfield.accessibility_tree).

`update_tree()` creates a new [`AccessibilityUpdate`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityUpdate.html) to track data relating to the update pass, updates its [`root_node_id`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#structfield.root_node_id) if necessary, and then updates itself based on data from the DOM.

### `AccessibilityUpdate`

> Methods in `AccessibilityTree` which can mutate the tree, such as [`get_or_create_node()`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#method.get_or_create_node) and `remove_stale_nodes()`, take an `AccessibilityUpdate` in order to track what changes were made.

Author: `remove_stale_nodes()` has been renamed? or will be renamed at least

Reviewer: Fixed.

If an `AccessibilityNode` is changed in any way, it is added to the `AccessibilityUpdate`'s [`changed_nodes`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityUpdate.html#structfield.changed_nodes) set.
This set is used to produce the `accesskit::TreeUpdate` when the `AccessibilityUpdate` is [finalized](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityUpdate.html#method.finalize).

### Initial tree construction

The first time [`update_tree`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#method.update_tree) is called on the accessibility tree, its [`nodes`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#structfield.nodes) map is empty and needs to be populated from the DOM tree.

> This will also produce an initial `accesskit::TreeUpdate` to populate the tree on the AccessKit side

Author: ```suggestion
This will produce an initial `accesskit::TreeUpdate` to populate the tree on the AccessKit side.
```

Reviewer: Done.

This method will then call into [`update_node_and_descendants_from_dom_node()`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#method.update_node_and_descendants_from_dom_node) for the root node.

[`update_node_and_descendants_from_dom_node()`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#method.update_node_and_descendants_from_dom_node) populates the root node in two steps:
- First [`AccessibilityNode::update_node_from_dom_node()`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityNode.html#method.update_node_from_dom_node) populates the node's local properties, such as its role.
- Second, [`AccessibilityNode::update_descendants_from_dom_node()`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityNode.html#method.update_descendants_from_dom_node) populates the node's children - recursively calling into `AccessibilityTree::update_node_and_descendants_from_dom_node()` for each new child.

> As the tree is populated from the DOM tree via `update_node_and_descendants_from_dom_node()`, the `AccessibilityUpdate` keeps track of changes to accessibility nodes which need to be resolved as updates to properties computed within the accessibility tree.

Author: lead with “Some properties in the accessibility tree, in particular a node’s computed name, are dependent on other information in the accessibility tree, so they need to be computed later.”

then explain what “local” means, then jump into how the damage works

Reviewer: Done.

This method first calls into [`AccessibilityNode::update_node_local()`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityNode.html#method.update_node_local) to resolve the local damage on the node, and then recurses into the node's children if necessary.

Finally, after these three update phases (creating the root node, populating it recursively from the DOM tree, and computing properties which are based on the populated accessibility tree) are complete, the `AccessibilityUpdate` is [`finalize()`d](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityUpdate.html#method.finalize), producing the initial `accesskit::TreeUpdate` for the tree.

### Incremental updates

Calls to [`update_tree()`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#method.update_tree) after the tree has been populated for the first time follow the same basic flow, but with differences resulting from the tree already being populated.

> Typically, calls to `ensure_root_node()` after the tree has been populated will be a mostly no-op, since there will already be a node associated with the root DOM node.

Author: ```suggestion
Typically, calls to `ensure_root_node()` after the tree has been populated will essentially be a no-op, since there will already be a node associated with the root DOM node.
```

Reviewer: Done.

The `damage_from_dom` argument to `update_tree()` should now contain pairs of corresponding `ServoLayoutNode`s and `AccessibilityDamage` values.
These are collected in between reflows on the Document's [`AccessibilityData`](https://doc.servo.org/script/dom/document/accessibility_data/struct.AccessibilityData.html), and passed in to the reflow method via [`ReflowRequest`](https://doc.servo.org/layout_api/struct.ReflowRequest.html#structfield.accessibility_damage).

The `damage_from_dom` structure is passed (as previously) to [`apply_changes_from_dom_tree()`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#method.apply_changes_from_dom_tree), where it now allows us to update only nodes whose corresponding DOM nodes have changed since the last tree update.

> Any nodes which have been added to the DOM tree will be picked up due to a parent node having a [`Children`](https://doc.servo.org/layout_api/struct.AccessibilityDamage.html#associatedconstant.Children) damage value; from there, any new nodes will be populated recursively the same way the root node is in the initial update.

Author: ```suggestion
Any nodes which have been added to the DOM tree since the last accessibility update will be picked up due to a parent node having a [`Children`](https://doc.servo.org/layout_api/struct.AccessibilityDamage.html#associatedconstant.Children) damage value; from there, any new nodes will be populated recursively the same way the root node is in the initial update.
```

Reviewer: Done.

> ### Dropping nodes from the cache: `TreeChange`

Author: ```suggestion
### Dropping nodes from the `AccessibilityTree`: `TreeChange`
```

Reviewer: Done.

Once the tree update is complete, any `AccessibilityNode` which is no longer in the tree is dropped.

> This happens in the [`AccessibilityTree::drop_removed_nodes()`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#method.drop_removed_nodes) method, which consumes the `AccessibilityUpdate`.

Author: ```suggestion
This happens in the [`AccessibilityTree::drop_removed_nodes()`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#method.drop_removed_nodes) method, which consumes parts of the `AccessibilityUpdate` during `finalize()`.
```

Reviewer: Done.

Author: possibly missing “during `finalize()`”, but i think that’s ok

Reviewer: Ah, guess I made an editing decision there. I think it's better with, so I added it back in.

`drop_removed_nodes()` is called at the end of the update so that we can distinguish between a node which has been removed from its parent node because it was moved elsewhere in the tree, and a node which has been removed from the tree altogether.

We determine which nodes to drop using the `AccessibilityUpdate`'s [`tree_changes`](`https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityUpdate.html#structfield.tree_changes`) map, which tracks what nodes have been added to, removed from and moved within the tree during the current update pass.

Tree changes are determined based on updates to parent nodes:
- If a DOM node has a child which hasn't yet been added to the tree, that node will be created when [`get_or_create_node()`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#method.get_or_create_node) is called, and it will be tracked as [`TreeChange::New`](https://doc.servo.org/layout/accessibility_tree/enum.TreeChange.html#variant.New)

>     - A node which is `New` a particular update can never be `Moved` or `Removed` in the same update.

Author: ```suggestion
    - A node which is `New` in a particular update can never be `Moved` or `Removed` in the same update.
```

Reviewer: Done.

- If a DOM node has a child which was previously the child of a different node, the child node will be tracked as [`TreeChange::PendingMove`](https://doc.servo.org/layout/accessibility_tree/enum.TreeChange.html#variant.PendingMove)

>     - If it was previously tracked as `Removed`, it will be tracked as `Moved`

Author: ```suggestion
- If a DOM node has a child which was previously the child of a different node, the child node will be tracked as [`TreeChange::PendingMove`](https://doc.servo.org/layout/accessibility_tree/enum.TreeChange.html#variant.PendingMove), or `Moved` if it was previously tracked as `Removed`.
```

Reviewer: Done.

- If a DOM node *no longer* has a child which it had previously, that child node will be tracked as [`TreeChange::Removed`](https://doc.servo.org/layout/accessibility_tree/enum.TreeChange.html#variant.Removed)

>     - If it was previously tracked as `PendingMove`, it will be tracked as `Moved`.

Author: ```suggestion
- If a DOM node *no longer* has a child which it had previously, that child node will be tracked as [`TreeChange::Removed`](https://doc.servo.org/layout/accessibility_tree/enum.TreeChange.html#variant.Removed), or `Moved` if it was previously tracked as `PendingMove`.
```

Reviewer: Done.

For a node to be tracked as [`Moved`](https://doc.servo.org/layout/accessibility_tree/enum.TreeChange.html#variant.Moved), it must be both removed from its old parent and added to its new parent in the same update pass, in either order.

> When [tree_changes](`https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityUpdate.html#structfield.tree_changes`) is processed by [`AccessibilityTree::drop_removed_nodes()`](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#method.drop_removed_nodes) during [finalization](https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityUpdate.html#method.finalize), any nodes in the subtree of a node with `Removed` status will be dropped from the tree's cache.

Author: do we define cache? maybe we can instead link to <https://doc.servo.org/layout/accessibility_tree/struct.AccessibilityTree.html#structfield.nodes>?

Reviewer: Done.
